staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"fmt"
+	"net/http"
+	"runtime"
+
+	// TODO: decide whether to also generate the old metrics, which
+	// categorize according to mutating vs readonly.
+
+	// "k8s.io/apiserver/pkg/endpoints/metrics"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	utilflowcontrol "k8s.io/apiserver/pkg/util/flowcontrol"
+	utilfilterconfig "k8s.io/apiserver/pkg/util/flowcontrol/filterconfig"
+	"k8s.io/klog"
+)
+
+// WithPriorityAndFairness limits the number of in-flight
+// requests in a fine-grained way.
+func WithPriorityAndFairness(
+	handler http.Handler,
+	longRunningRequestCheck apirequest.LongRunningRequestCheck,
+	fcIfc utilflowcontrol.Interface,
+) http.Handler {
+	if fcIfc == nil {
+		klog.Warningf("priority and fairness support not found, skipping")
+		return handler
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no RequestInfo found in context"))
+			return
+		}
+		user, ok := apirequest.UserFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no User found in context"))
+			return
+		}
+		requestDigest := utilfilterconfig.RequestDigest{requestInfo, user}
+
+		// Skip tracking long running requests.
+		if longRunningRequestCheck != nil && longRunningRequestCheck(r, requestInfo) {
+			klog.V(6).Infof("Serving RequestInfo=%#+v, user.Info=%#+v as longrunning\n", requestInfo, user)
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		execute, afterExecute := fcIfc.Wait(ctx, requestDigest)
+		if !execute {
+			tooManyRequests(r, w)
+			return
+		}
+		defer afterExecute()
+
+		// Serve the request, but asynchronously, and return from here
+		// as soon as either the request is finished or the context is
+		// canceled.  The logic here is heavily cribbed from the
+		// timeout filter.
+		timedOut := ctx.Done()
+
+		// gets sent to exactly once, with the result of recover() for serving the request
+		resultCh := make(chan interface{})
+
+		go func() { // serve the request, with recovery from panics
+			defer func() {
+				err := recover()
+				if err != nil && err != http.ErrAbortHandler {
+					// Same as stdlib http server code. Manually allocate stack
```

Comment:
I think in at least one other place we actually just send whatever this is through a channel and re-panic, to be sure that the ordinary panic handler gets it. (I don't think I'm asking for a change right now)

Reply:
That actually is what happens here (as well as in the timeout handler). I think the extra stack trace gets logged so that the original panic's trace doesn't get lost; sadly, I don't think Go has a Java-like way to capture a panic's original stack trace in a non-lossy way for repanicking later.

(Even more sadly, the fact that this is also done in the timeout controller, and possibly other places too, means that the ultimate panic is going to have lots of stack traces on it, which kind of sucks.)

Reply:
Also in Java it happens that there are multiple printouts of various vintages of a problematic stack.